Allow zipped ranges to have different sizes

romainbrenguier opened the pull request:
This adds a template argument with which we can allow ranges to have different sizes, in which case the end of the zip result is defined by the first of the two ranges ending.

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [na] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [na] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [na] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

romainbrenguier committed:
Allow zipped ranges to not have the same size

This is parameterized by a template argument.

romainbrenguier committed:
Add unit test zipping ranges of different sizes

smowton commented on src/util/range.h:
```diff
@@ -283,6 +288,9 @@ struct zip_iteratort
 
   bool operator==(const zip_iteratort &other) const
   {
+    if(!same_size && end_reached() == other.end_reached())
```
Surely `end_reached() && other.end_reached()`

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 6185c2a7e27f7de1076c625a1f99bb6470c8e7b7).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/115814656

smowton reviewed (changes requested):
Also this suggests iterator equality might not be tested?

romainbrenguier commented:
@smowton 
> Also this suggests iterator equality might not be tested?

It was only exercised with the second iterator being the end iterator. I've added a check to the unit test which would fail with the buggy version of equality (here: https://github.com/diffblue/cbmc/pull/4807/files#diff-0f891f8544182f8d344c5e6aeb9fede2R265)

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 799360d619adadb323fd60ee3918acde831dfe15).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/115835286

thomasspriggs commented on src/util/range.h:
```diff
@@ -347,6 +358,19 @@ struct zip_iteratort
   second_iteratort second_end;
   std::shared_ptr<value_type> current = nullptr;
+
+  bool end_reached() const
```
:question: Now that you have this separate `end_reached` function, could the following invariant in the increment operator be simplified?
```
PRECONDITION(
      !same_size ||
      ((first_begin == first_end) == (second_begin == second_end)));
```
becomes
```
   PRECONDITION(!end_reached());
```

thomasspriggs commented on src/util/range.h:
```diff
@@ -421,23 +445,23 @@ struct ranget final
 
-  template <typename other_iteratort>
-  ranget<zip_iteratort<iteratort, other_iteratort>>
+  template <bool same_size = true, typename other_iteratort>
```
:pick: Missing doxygen.

thomasspriggs commented on src/util/range.h:
```diff
@@ -421,23 +445,23 @@ struct ranget final
   }
 
-  template <typename containert>
+  template <bool same_size = true, typename containert>
```
:pick: Missing doxygen.

thomasspriggs commented on src/util/range.h:
```diff
@@ -299,7 +307,8 @@ struct zip_iteratort
     INVARIANT(
-      (first_begin == first_end) == (second_begin == second_end),
+      !same_size ||
+        ((first_begin == first_end) == (second_begin == second_end)),
```
:bulb: Add an instance of `cbmc_invariants_should_throwt` to your unit test and then test that the invariant is violated when you attempt to increment the iterator beyond the end for various combinations of

- First collection shorter.
- Second collection shorter.
- Both collections the same length with `is_same` set to `true` and `is_same` set to `false`.

It is getting more difficult to see that the invariants are correct by inspection. Therefore it would be good to unit test them.

thomasspriggs commented on unit/util/range.cpp:
```diff
@@ -245,6 +245,39 @@ SCENARIO("range tests", "[core][util][range]")
+    const std::vector<std::string> string_vector{"foo", "bar", "baz", "bay"};
+    WHEN("We zip the vectors")
+    {
+      auto range = make_range(int_vector).zip<false>(string_vector);
```
:pick: Please repeat this test, but with `int_vector` and `string_vector` swapped.

thomasspriggs reviewed (commented):
I think this could really use a few more unit tests, before we merge it and start basing too many things on this functionality.

romainbrenguier committed:
Simplify operator++ precondition using end_reached

Now that this method is there, it makes the precondition simpler.

romainbrenguier committed:
Document zip method

The meaning of the first template argument was not obvious.

romainbrenguier committed:
Check for invariant throw in unit tests

This check that a zipped range will throw invariant failures in case
same_size is true and we reach one of the two ends but not the other.

romainbrenguier commented:
@thomasspriggs I added more to the unit tests

thomasspriggs reviewed (approved):
Nice

smowton commented on unit/util/range.cpp:
```diff
@@ -245,6 +245,126 @@ SCENARIO("range tests", "[core][util][range]")
+        REQUIRE(second_pair.first == 51);
+        REQUIRE(second_pair.second == "bar");
+      }
+      second_range = std::move(second_range).drop(1);
```
`third_range` for symmetry?

smowton commented on unit/util/range.cpp:
```diff
@@ -245,6 +245,126 @@ SCENARIO("range tests", "[core][util][range]")
+        REQUIRE(second_pair.first == "bar");
+        REQUIRE(second_pair.second == 51);
+      }
+      second_range = std::move(second_range).drop(1);
```
third_range as above

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: 3e2a5d3658dba97e32db0b9a04ff3e385769fa9a).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/116296832
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: 8fafc74db7dac34862da6516b329717225c84022).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/116318287
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 9faecc0eee8bea4b1672ee13c8efd20206bfd336).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/116321408

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 5afafe04ef3622aba5fb8d3e372abd203cd17325).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/116391281
